tools/nebula-flink/src/main/java/org.apache.flink/connector/nebula/source/NebulaInputFormat.java

```diff
@@ -0,0 +1,162 @@
+        }
+        rows = new ArrayList<>();
+        formatConverter = new NebulaRowInputFormatConverter();
+    }
+
+    @Override
+    public void closeInputFormat() throws IOException {
+        try {
+            if (client != null) {
+                connectionProvider.close();
+            }
+        } catch (Exception e) {
+            LOG.error("close client error,", e);
+            throw new IOException("close client error,", e);
+        }
+    }
+
+    @Override
+    public BaseStatistics getStatistics(BaseStatistics baseStatistics) throws IOException {
+        return baseStatistics;
+    }
+
+    @Override
+    public InputSplit[] createInputSplits(int numSplit) throws IOException {
+        return new GenericInputSplit[]{new GenericInputSplit(0,1)};
+    }
+
+    @Override
+    public InputSplitAssigner getInputSplitAssigner(InputSplit[] inputSplits) {
+        return new DefaultInputSplitAssigner(inputSplits);
+    }
+
+    @Override
+    public void open(InputSplit inputSplit) throws IOException {
+        if (inputSplit != null) {
+            Map<String, List<String>> returnCols = new HashMap<>();
+            returnCols.put(executionOptions.getLabel(), executionOptions.getFields());
+
+            Processor processor;
+            Iterator iterator;
+            if (executionOptions.getDataType().isVertex()) {
+                iterator = storageClient.scanVertex(executionOptions.getGraphSpace(), returnCols, executionOptions.isAllCols(), executionOptions.getLimit(), executionOptions.getStartTime(), executionOptions.getEndTime());
+                processor = new ScanVertexProcessor(metaClient);
+            } else {
+                iterator = storageClient.scanEdge(executionOptions.getGraphSpace(), returnCols, executionOptions.isAllCols(), executionOptions.getLimit(), executionOptions.getStartTime(), executionOptions.getEndTime());
+                processor = new ScanEdgeProcessor(metaClient);
+            }
+
+            while (iterator.hasNext()) {
+                Result result = processor.process(executionOptions.getGraphSpace(), iterator.next());
+                List<com.vesoft.nebula.data.Row> dataset = result.getRows(executionOptions.getLabel());
+                rows.addAll(dataset);
+            }
+            resultIter = rows.iterator();
+            hasNext = resultIter.hasNext();
+        }
+    }
+
+    @Override
+    public boolean reachedEnd() throws IOException {
+        return !hasNext;
+    }
+
+    @Override
+    public Row nextRecord(Row reuse) throws IOException {
+        if (!hasNext) {
+            return null;
+        }
+        scannedRows++;
+        Row row = (Row) formatConverter.convert(resultIter.next(), executionOptions.getDataType().isVertex());
+        hasNext = resultIter.hasNext();
+        return row;
+    }
+
+    @Override
+    public void close() throws IOException {
+        LOG.info("Closing split (scanned {} rows)", scannedRows);
```

Comment:
close client

Reply:
client is closed in NebulaSourceFunction.close